### PR TITLE
Add http servers max body size and max headers size limits

### LINKED ISSUE #3372: Configurable body size
I tried to change the default body size via undertow-handlers.conf but it's not working.

The undertow-handlers.conf file contains the following :

`method(POST) -> buffer-request(buffers=20000000)`

Also, @gsmet suggested that it would be better expose this kind of configuration in application.properties

P.S. This is the first time that i use undertow so i might got the configuration wrong.

### PR DESCRIPTION
**Pull request by machi1990**
Fixes #3372

### CODE REVIEW

**Comment by gsmet** [comment]
Looks like the build is failing.

**Comment by mkouba** [comment]
```
[ERROR] Errors: 
[ERROR]   DevMojoIT.testThatApplicationRecoversStartupIssue:430->runAndExpectError:561->MojoTestBase.getHttpErrorResponse:196 » ConditionTimeout
```

**Comment by machi1990** [comment]
@gsmet @mkouba was a bit busy today, hoping to debug this failure tonight.

File: docs/src/main/asciidoc/undertow-reference.adoc [hunk 1]
```diff
@@ -143,3 +143,27 @@ quarkus.http.cors.exposed-headers=Content-Disposition
...
+It is possible to provide some options that control server behaviour. The following properties are supported.
+[cols="<m,<m,<2",options="header"]
+|===
+|Property Name|Default|Description
+|quarkus.http.server-options.max-headers|`200`| The maximum number of headers that are permitted in a request.
+|quarkus.http.server-options.max-parameters|`1000`|The maximum number of query parameters that are permitted in a request.
+|quarkus.http.server-options.max-header-size|`50K`|The maximum size of a HTTP header block.
+|quarkus.http.server-options.max-buffered-request-size|`16384`|The maximum size of a request that can be saved.
+|quarkus.http.server-options.max-entity-size| `unlimited` |The default maximum size of a request entity.
+|quarkus.http.server-options.multipart-max-entity-size| `unlimited` | The default max entity size when using the Multipart parser.
```
Comment (gsmet) [inline docs/src/main/asciidoc/undertow-reference.adoc]: As they are all related to `limits`, maybe I would have made them `quarkus.http.limits.max-entity-size`.

Just a suggestion though, wait for other opinions before changing.

Comment (machi1990) [reply docs/src/main/asciidoc/undertow-reference.adoc]: I have been thinking about this, I am sold to the idea. So let's do this.

**Comment by machi1990** [comment]
@stuartwdouglas @geoand would one of you guys like to review this PR? thanks.

**Comment by geoand** [comment]
@machi1990 thanks for the PR! @stuartwdouglas is the SME on this so best he does the review. 
That said I will go over it and see if I spot anything

**Comment by stuartwdouglas** [comment]
There are big changes coming to the Undertow subsystem after the next release that these will conflict with. See https://github.com/quarkusio/quarkus/compare/master...stuartwdouglas:vertx-new for details

**Comment by machi1990** [comment]
> There are big changes coming to the Undertow subsystem after the next release that these will conflict with. See [master...stuartwdouglas:vertx-new](https://github.com/quarkusio/quarkus/compare/master...stuartwdouglas:vertx-new) for details

@stuartwdouglas Good to know about these incoming changes. Do you think that the enhancement can still be done this way ? Will it make sense to contribute a PR directly to the changes so that it will land with the changes in the undertow subsystem?

**Comment by gsmet** [comment]
@machi1990 I would say let's wait for Stuart's PR as it has a high priority for us. Then we can adjust this.

File: bom/runtime/pom.xml [hunk 2]
```diff
@@ -852,12 +852,6 @@
                 </exclusions>
                 <version>${tika.version}</version>
             </dependency>
-            <dependency>
-                <groupId>io.quarkus.http</groupId>
-                <artifactId>quarkus-http-core</artifactId>
-                <version>${quarkus-http.version}</version>
-            </dependency>
```
[the remark this answers is not in the document]
Comment (machi1990) [reply bom/runtime/pom.xml]: It is not needed anymore. I have a rebase to make.

**Comment by gsmet** [comment]
@machi1990 there's a big issue with the rebase. It brought a lot of unrelated commits.

**Comment by machi1990** [comment]
> @machi1990 there's a big issue with the rebase. It brought a lot of unrelated commits.

Oops, I missed that. For some reason, I'll start over the rebase.

**Comment by machi1990** [comment]
I have rebased this PR after the split of vertx landed. @stuartwdouglas @gsmet @cescoffier  would any of you have some time for a review? Thanks.

File: extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowDeploymentRecorder.java [hunk 3]
```diff
@@ -327,6 +327,8 @@ public void run() {
             @Override
             public void handle(HttpServerRequest event) {
                 VertxHttpExchange exchange = new VertxHttpExchange(event, ALLOCATOR, executorService);
+                httpConfiguration.limits.maxBodySize
```
Comment (stuartwdouglas) [inline extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowDeploymentRecorder.java]: Can this be changed to just a .isPresent() call? Lambdas use a significant amount of metaspace and result in an allocation on every request, so we try and avoid them in the runtime code as much as possible.

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ServerLimitsConfig.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.configuration.MemorySize;
+
+@ConfigGroup
+public class ServerLimitsConfig {
+    /**
+     * The the maximum length of all headers.
+     */
+    @ConfigItem(defaultValue = "8192")
+    public MemorySize maxHeaderSize;
```
Comment (stuartwdouglas) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ServerLimitsConfig.java]: I am not sure if this is big enough as a default, some sites have a lot of cookies, even for this page on github my browser is sending 1.4k of headers.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
Other than those two minor comments looks good

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ServerLimitsConfig.java (hunk 4 above)
Comment (machi1990) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/ServerLimitsConfig.java]: I'll update the default value to something like `20K`.

File: extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowDeploymentRecorder.java (hunk 3 above)
Comment (machi1990) [reply extensions/undertow/runtime/src/main/java/io/quarkus/undertow/runtime/UndertowDeploymentRecorder.java]: > Other than those two minor comments looks good

Thanks for the review @stuartwdouglas. New updates pushed. WDYT?

**Review by stuartwdouglas — APPROVED** [review]

**Pull request merged**